### DISCUSSION TITLE
Equivalent Capacitance/Inductance

### DISCUSSION THREAD

**Discussion by BEAV-SahagunJ** [SpiceSharp/SpiceSharp#163] [Q&A] [answered]
Hi Sven,

I know this might be another "noob" question I'm about to ask. So I want to apologize in the beginning of this post. Pardon me as I am trying to understand this even though my degree is far from Electrical Engineering. So, here goes nothing.

I found a document which gets the Equivalent Resistance in PSPICE. See below for reference:

![image](https://user-images.githubusercontent.com/83271157/118226260-340abb00-b4b9-11eb-9c07-ca241ac66848.png)

From that reference, I managed to make it work using your great library (by the way, this is for a thesis project).

So my question, **Is there any same implementation for Equivalent Capacitance/Inductance in SpiceSharp?** I think there is, I just don't what to use and how to use it. 

Please don't get me wrong, I am trying my best to understand and utilize every function I can use. Just don't know which one this time. :)

Thanks for your usual support!

**Comment by svenboulanger** [comment, marked as answer]
Hi there!

Computing an equivalent capacitance/inductance is possible and relatively similar unless you decide to mix it up (ie RLC networks, or even harder, nonlinear stuff).

For capacitors, you can apply a constant (DC) current + initial condition (or a pulsed current source), and measure the voltage slope of the charging capacitor circuit while the current is constant in a transient simulation. 

For inductors, you can apply a pulsed voltage source and measure the slope of the voltage  source current (or again use initial conditions). 

In both cases, the information you're looking for is time-dependent, so just a DC or OP simulation won't cut it. 

There are a few alternatives:
- You can also avoid both methods entirely and do an AC simulation. The so-called impedance or admittance then directly reflects the equivalent capacitance/inductance.
- you can do a transient simulation with a sine wave to get similar information as with an AC simulation. The amplitude and phase of the measurement should be the same as an AC simulation. 

In my opinion, the AC simulation will be the fastest, but it may not be the best method to learn what's going on. ;-)

**Reply by BEAV-SahagunJ** [reply]
Hi again,

Thanks for your answer. Much appreciated. Another thing, can you give me an example? A scratch maybe where I can start? Thanks!

Sorry for this shallow questions :(

**Reply by BEAV-SahagunJ** [reply]
Is your suggestion on my scenario? I am receiving voltage loop rule violation in my circuit. Am I missing something? My circuit only contains 3 capacitors and 1 voltage source with 1volts. Hope I get a reply. Need it badly.

**Reply by marcin-golebiowski** [reply]
Please share your code if possible, then I will try to help you.

**Reply by BEAV-SahagunJ** [reply]
```
Circuit ckt = new Circuit(
    new VoltageSource("V1", "1", "0", 1),
    new Capacitor("C1", "1", "2", 1.0E-6),
    new Capacitor("C2", "2", "3", 1.0E-6),
    new Capacitor("C3", "3", "0", 1.0E-6)
    );

AC ac = new AC("AC 1", new DecadeSweep(1, 1, 1));

ac.ExportSimulationData += (sender, e) =>
{

};

ac.Run(ckt); // Voltage Loop Violation.
```

This is what I have. Just testing for now. Let me know what should I do next.

**Reply by BEAV-SahagunJ** [reply]
To give you guys an idea of what I am working. It is just a simple LCR Meter.

For Example, the user connects 3 Resistors with different values in series, parallel or both, then the LCR will give the total resistance.

Same with Capacitor and Inductor. I am not into much complex RLC network. All I have is 1 Component Type every reading. So I think it is not that complex. I just need to show the result every I connect the LCR meter to point A and B just like the image on my post. 

I was able to do it for Total Resistance using just the DC simulation because I just added dummy Voltage Source with 1V and used Rt = Vdc/Idc.

Just help me with the last 2. Capacitance and Inductance.

Hope you guys can help me. And sorry if I'm to noob for this. I'm trying to learn. Thanks!

**Reply by marcin-golebiowski** [reply]
It looks like there is a bug in Spice# when connecting the capacitors in series in AC simulation.
I'm trying to help you, but I'm having some difficulties to write something useful to you. I will get back you as soon as possible.

**Reply by marcin-golebiowski** [reply]
You can compute capacitance of network by using CR network and measure the tau of the network.

```
using SpiceSharp;
using SpiceSharp.Components;
using SpiceSharp.Simulations;
using System;

namespace ConsoleApp
{
    class Program
    {
        static void Main(string[] args)
        {
            double dcVoltage = 10;
            var resistorResistance = 1000;
            var capacitance = 1e-3;

            // Build circuit
            var ckt = new Circuit(
                new VoltageSource("V1", "IN", "0", dcVoltage),
                new Resistor("R1", "IN", "OUT", resistorResistance),
                new Capacitor("C1", "OUT", "A_1", capacitance * 2),
                new Capacitor("C2", "A_1", "0", capacitance * 2)
            );

            var tran = new Transient("tran", 1e-4, 5, 1e-4);
            tran.TimeParameters.InitialConditions["OUT"] = 0.0;
            tran.TimeParameters.InitialConditions["A_1"] = 0.0;

            IExport<double> export = new RealCurrentExport(tran, "V1");

            double prevCurrent = 0.0;
            double? lastTime = null;
            double? initialCurrent = null;

            tran.ExportSimulationData += (sender, data) =>
            {
                var current = export.Value;

                if (initialCurrent == null)
                {
                    initialCurrent = current;
                }

                var currentAbs = Math.Abs(current);

                if (lastTime == null && currentAbs <= 0.00673794699 * Math.Abs(initialCurrent.Value)) // 5 Tau => 0.00673794699 of initial current
                {
                    lastTime = data.Time;
                }
                prevCurrent = current;

            };
            tran.Run(ckt);

            var tau = lastTime / 5.0; // compute tau

            var capacitanceCalculated = tau / resistorResistance;

            Console.WriteLine("Capacitance calculated = " + capacitanceCalculated + ", should be = " + capacitance);
        }
    }
}
```

**Reply by marcin-golebiowski** [reply]
Please find https://www.electronics-tutorials.ws/rc/rc_1.html article for information.

**Reply by svenboulanger** [reply]
> ```
> Circuit ckt = new Circuit(
>     new VoltageSource("V1", "1", "0", 1),
>     new Capacitor("C1", "1", "2", 1.0E-6),
>     new Capacitor("C2", "2", "3", 1.0E-6),
>     new Capacitor("C3", "3", "0", 1.0E-6)
>     );
> 
> AC ac = new AC("AC 1", new DecadeSweep(1, 1, 1));
> 
> ac.ExportSimulationData += (sender, e) =>
> {
> 
> };
> 
> ac.Run(ckt); // Voltage Loop Violation.
> ```
> 
> This is what I have. Just testing for now. Let me know what should I do next.

This should result in a floating node rule violation, not a voltage loop one... I advise either adding initial conditions to any floating nodes if you go the transient route, or have a very large resistor to ground from each floating node (this is what most commercial Spice simulators do while giving a warning). You can check which nodes are floating from the rule violation.

If you plan on physically making an LCR meter, then an ideal current source would be difficult.
IIRC, most "simple" LCR meters will apply a sine wave or other known waveform, and then calculate the resistor and inductance/capacitance from the amplitude and phase response on the waveform.

This should work (and gives me ~0.333uF as expected, there is a very small error due to the "fix" resistors).

```csharp
Circuit ckt = new Circuit(
    new VoltageSource("V1", "1", "0", 1)
        .SetParameter("acmag", 1.0),
    new Capacitor("C1", "1", "2", 1.0E-6),
    new Capacitor("C2", "2", "3", 1.0E-6),
    new Capacitor("C3", "3", "0", 1.0E-6),

    // Fix floating nodes
    new Resistor("Rfix1", "2", "0", 1e12),
    new Resistor("Rfix2", "3", "0", 1e12)
    );

double frequency = 1.0 / (2 * Math.PI); // f (Hz) = 1 (rad) / (2pi)
AC ac = new AC("AC 1", new DecadeSweep(frequency, frequency, 1));

var current = new ComplexCurrentExport(ac, "V1");
ac.ExportSimulationData += (sender, e) =>
{
    Console.WriteLine(-current.Value.Imaginary);
};

ac.Run(ckt);
```

**Reply by BEAV-SahagunJ** [reply]
Hi sven,

Thanks for this. It works. I also tried it in parallel and series-parallel combinations. Yes, there's a small difference from the expected result. Is there any additional fix I can apply to reduce the difference by atleast +-1%?

The difference is somehow visible for parallel and series-parallel combinations only. Series looks good.

**Reply by svenboulanger** [reply]
That is relatively straight-forward actually.
Just do the AC simulation at a higher frequency, and scale it back down afterwards.

```csharp
Circuit ckt = new Circuit(
    new VoltageSource("V1", "1", "0", 1)
        .SetParameter("acmag", 1.0),
    new Capacitor("C1", "1", "2", 1.0E-6),
    new Capacitor("C2", "2", "3", 1.0E-6),
    new Capacitor("C3", "3", "0", 1.0E-6),

    // Fix floating nodes
    new Resistor("Rfix1", "2", "0", 1e12),
    new Resistor("Rfix2", "3", "0", 1e12)
    );

double frequency = 1000.0 / (2 * Math.PI); // f (Hz) = 1 (rad) / (2pi)
AC ac = new AC("AC 1", new DecadeSweep(frequency, frequency, 1));

var current = new ComplexCurrentExport(ac, "V1");
ac.ExportSimulationData += (sender, e) =>
{
    Console.WriteLine(-current.Value.Imaginary / 1000);
};

ac.Run(ckt); // Voltage Loop Violation.
```

**Reply by BEAV-SahagunJ** [reply]
Thanks. Got it working. One last question. I'll take my chances here. I'm almost finish, big big thanks to you guys especially sven.

Does this code also applies to Inductance? Or do I need to change some parameter?

Tried to run the same code (just replacing the Capacitor with Inductor) which resulted to VoltageLoopRule.

**Reply by BEAV-SahagunJ** [reply]
Hi there,

I managed to make it work on inductance for Series and the result is correct with expected. I just added 1 Resistor with 0 ohms. But for Parallel and Series-Parallel connection, it doesn't work. I keep on getting Voltage Loop and Floating Node Violation. 

```
Circuit ckt = new Circuit(
    new VoltageSource("V1", "1", "0", 1)
        .SetParameter("acmag", 1.0),
    new Inductor("L1", "1", "0", 1.0E-3),
    new Inductor("L2", "1", "0", 1.0E-3),
    new Inductor("L3", "1", "0", 1.0E-3),

    new Resistor("R1", "1", "0", 0),
    );

double frequency = 1.0 / (2 * Math.PI); // f (Hz) = 1 (rad) / (2pi)
AC ac = new AC("AC 1", new DecadeSweep(frequency, frequency, 1));

var current = new ComplexCurrentExport(ac, "V1");
ac.ExportSimulationData += (sender, e) =>
{
    Console.WriteLine(1 / current.Value.Imaginary);
};

ac.Run(ckt);
```
What am I missing?

**Reply by svenboulanger** [reply]
Inductors are basically short-circuits at DC. If you put inductors in parallel, it can't compute how much current is going through each, because they act like shorts (a short is like a 0V voltage source, hence the voltage loop error).

Add series resistors with each inductor and make it very small to avoid the issue.

**Reply by BEAV-SahagunJ** [reply]
Resistor on each side? Sorry if its not clear to me. Can you give me an example? This is the last function I need to make it work and I'm done with this project. Please please.

**Reply by svenboulanger** [reply]
One side, both sides, whatever you want man, it all works.
